src/rows.jl

```diff
@@ -199,6 +202,16 @@ Row(schema::Schema, fields) = Row(schema, NamedTuple(Tables.Row(fields)))
 Row(schema::Schema, fields::Row) = Row(schema, getfield(fields, :fields))
 Row(schema::Schema, fields::NamedTuple) = Row(schema; fields...)
 
+"""
+    Legolas.Schema(row::Row)
+    Legolas.Schema(::Type{<:Row})
+
+Return the schema for the given [`Legolas.Row`](@ref) or row type.
+"""
+Schema(row::Row) = Schema(typeof(row))
```

Comment:
I’m mostly fine with this but I wonder if it should be a separate function rather than a constructor

the reason being that it encourages writing code that naive authors might think is “generic” but isn’t as generic as most producer/consumer code probably should be (ie not caring whether you’re given an actual Legolas.Row or just a compliant Tables.jl row)

actually, maybe that brings up a good point - can you point me to code where this mechanism is useful in practice? I wonder whether the need here is actually pointing out an upstream code smell or something (since it implies the existence of code that “cares” if it’s being given a Legolas.Row, without that code itself being the source of the Legolas.Row)

(keep in mind this is OSS, so if internal Beacon code, can link on Slack)

Reply:
It's internal code but the gist is that I often find myself writing `MyRowType = @row(...)`, applying `MyRowType` to some rows, then wanting to `write` out the result, but `write` requires a `Schema` and I only have the row type so I have to manually make a `Schema` type and type out the schema name again. It seems simpler and potentially less error prone to me to allow the user to use a row type when they have one. Then the call becomes `write(path, table, Schema(MyRowType))` (or just `MyRowType` with the `write` method change).

I guess a fairly direct OSS analogue is https://github.com/beacon-biosignals/Onda.jl/blob/master/src/signals.jl#L137, which requires rewriting the schema for `Signal`. When Onda bumps the `Signal` schema version, it'll have to be changed in multiple places, since it's written out explicitly rather than using the `Signal` type it already has.

Reply:
> MyRowType = @row(...), applying MyRowType to some rows, then wanting to write out the result, but write requires a Schema and I only have the row type so I have to manually make a Schema and type out the schema name again

I'm not sure I understand "and I only have the row type" - if you defined the row type, you should know the version of the schema you defined, no? In terms of changing it in multiple places - you'd have to do basically the same find-replace either way, right?

Does this come down to aesthetics for the "spelling of the invocation", or is there a different issue at play? The reason I'm inclined to think it's the former is because non-literal usages of this kind of path are a bit trickier in generic contexts for the aforementioned reason, so I'd assume it'd mostly get used "correctly" in the literal form, and in that case it essentially is an aesthetic choice

FWIW on the aesthetics front it makes more sense to me to steer table-level operations with `Schema` rather than row type constructors. Though I can see that `MyRowType` might read as less verbose than `Schema("my-schema@1")`

on the non-aesthetics front, the main two things i'd like to avoid/discourage for downstream users include:

- writing faux-generic code where behavior/correctness is predicated on whether or not a given row is a `Legolas.Row` vs. any compliant Tables.jl row 
- problematic assumptions/constraints on table types/parameters (e.g. a valid `Vector{NamedTuple{...}}` that can't be consumed properly because of assumptions about row type)

> I guess a fairly direct OSS analogue is https://github.com/beacon-biosignals/Onda.jl/blob/master/src/signals.jl#L137, which requires rewriting the schema for Signal. When Onda bumps the Signal schema version, it'll have to be changed in multiple places, since it's written out explicitly rather than using the Signal type it already has.

my preference is for these kinds of interfaces to be functional/value driven at the table level and not `Type{...}`-driven/row-driven for the above reasons

Reply:
> I'm not sure I understand "and I only have the row type" - if you defined the row type, you should know the version of the schema you defined, no?

Yes, you have the schema string, just not the `Schema` object you need in order to `write`. Constructing that requires rewriting the schema string instead of being able to query the row type for its schema (which would be particularly handy when you aren't the one who wrote the definition).

> In terms of changing it in multiple places - you'd have to do basically the same find-replace either way, right?

Perhaps I'm not understanding what you mean but with what I'm suggesting, you only ever write out the schema string once, so there's no find and replace to do other than modifying the version number at the definition site. So for example, Onda defines `Signal` and never has to write out `Schema("signal@1")` (or whatever its schema is) again; all further uses of `Signal` that need to know its schema can call `Schema(Signal)` or potentially use `Signal` itself directly.

> writing faux-generic code where behavior/correctness is predicated on whether or not a given row is a `Legolas.Row` vs. any compliant Tables.jl row

In this case, it's a strict addition of flexibility, not a change in behavior or correctness based on the input. The assumption is only that `Schema` can be called on whatever it is that was passed as that argument to `write`; users can still pass a `Schema` as before and it will behave the same, they just get the added convenience of being able to pass in things that have known schemas.

FWIW though I care less about the `write` method change than I do about being able to call `Schema` on rows/row types because the latter permits e.g. `write(path, table, Schema(MyRow))` which is good enough for me.

Reply:
> rewriting the schema string instead of being able to query the row type for its schema

This is what I think I mean by "this seems like a spelling/aesthetics thing": It seems like the difference is "you have to write `Schema("my-schema@1")`" vs. "you have to write `Schema(MyRowType)`". From a pure spelling perspective, I see no benefit to the latter and prefer the former

From a NOT pure spelling benefit (i.e. in generic code), I'm not sure I see any benefit but the downsides/increased surface for mistakes in generic code are pretty easy to imagine. 

Actually, there might be a better way to explain my reasoning/reluctance here: Let's assume we added this functionality, and walk through some hypothetical code review scenarios where it is used "in the wild."

- Scenario 1: I'm reviewing some code that invoked `Schema(T)` where `T` is a type parameter passed into the invocation context by an upstream caller. My review would be: "Let's refactor this to get rid of this invocation, and replace it with something else. Either `T` is a derived type parameter (e.g. an `eltype`) in which case this code is depending on things it shouldn't because its behavior will depend on whether an upstream type was inferred tightly enough (ref YASGuide's guidance on type parameter dependence), or it is a concrete type, in which case it's non-generic to other valid Tables.jl row types in a manner that's opaque to the caller"

- Scenario 2: I'm reviewing some code that invokes `Schema(row)` where `row` is a row value (in the Tables.jl sense) passed into the invocation context by an upstream caller. My review would be: "Let's refactor this to get rid of this invocation, and replace it with something else; this code now assumes that `row` is a `Legolas.Row`, which is an unnecessary constraint that is also likely to make usages of this code fragile/annoying in most contexts
 
 - Scenario 3: I'm reviewing some code that invokes `Schema(MyRowType)` where `MyRowType` is a literal type name/alias. My review would be: "Let's spell this as `Schema("my-schema@1")` instead, since it reads better and also surfaces the schema version at the callsite. Or, if we instead want the schema version to be driven by some upstream API function (that itself is hopefully SemVer'd), we can call that function instead."

Unless I'm missing something, these scenarios basically cover all usage cases for this functionality (please correct if I'm wrong!) and in every scenario my suggestion would be to refactor away the usage of this functionality; thus, my conclusion is that it means we probably shouldn't add this functionality

Does that make sense? or I might still be missing something interesting going on here

Reply:
>  there's more room for error if you need to remember and parrot the exact schema string and version rather than using an existing variable that stores it.

This is what I mean by it being a "spelling issue" - you're either "parroting" `"my-schema@1"` or "parroting" `MyRowType`. The typo mistake could happen just as easily either way.  The only difference in our scenario 3 is that a `MyMisspelledRowType` will error slightly earlier than `"my-schema@1"`. If scenarios 1/2 were valid, I'd be closer to agreeing with you, because in those scenarios the earlier errors could be "early enough" to actually be useful; but only looking at scenario 3, there's not a real benefit AFAICT. And actually, there's another class of errors to consider: that whatever literal is used is spelled correctly at the callsite, but that the prerequisite code hasn't been loaded. In that case, `"my-schema@1"` will result in a WAY more user-friendly error (since this case is so common, we added a special error path for it https://github.com/beacon-biosignals/Legolas.jl/pull/28).

Finally, even if we didn't align on the "spelling perspective" presented above, if you're seeking to use a constant binding to identify a schema rather than a literal string, then I'd say just do that instead of punning on a row type.

```
const MY_SCHEMA = Legolas.Schema("my-schema@1")
```

Though as mentioned previously, if we were to do this and thus remove callsite-specification of the schema version (not usually my preference), it's probably better to expose a functional interface to the hardcoded Schema (e.g. `do_with_my_schema(...)` or even just `my_schema()`) rather than a constant binding to it.

IIUC it seems like any of the above would give you any of the benefits you'd get with `Schema(::Type{<:Row})`, but without the downsides